Open Judge System/Web/OJS.Web/Areas/Administration/Controllers/SubmissionsController.cs

```diff
@@ -340,12 +341,25 @@ public ActionResult Retest(int id)
             }
             else
             {
-                if (!submission.ProblemId.HasValue || !this.CheckIfUserHasProblemPermissions(submission.ProblemId.Value))
+                bool isAdmin = this.User.IsAdmin();
+                if (isAdmin)
                 {
-                    this.TempData[GlobalConstants.DangerMessage] = "Нямате привилегиите за това действие";
-                    return this.RedirectToAction("Index", "Contests", new { area = "Administration" });
+                    if (!submission.ProblemId.HasValue ||
+                        !this.CheckIfUserHasProblemPermissions(submission.ProblemId.Value))
+                    {
+                        this.TempData[GlobalConstants.DangerMessage] = "Нямате привилегиите за това действие";
+                        return this.RedirectToAction("Index", "Contests", new {area = "Administration"});
+                    }
                 }
-
+                else
+                {
+                    if (!string.IsNullOrEmpty(submission.TestRunsCache) || !this.CheckIfUserOwnsSubmission(id))
```

Comment:
In the above if, there is a check for !submission.ProblemId.HasValue, but this check is missing here.